proposals/records/005-sharding.md

```diff
@@ -0,0 +1,197 @@
+produce a new segment. That new segment will work along with other existing
+segments to provide the shard's full dataset.
+
+[more details on the implementation provided elsewhere]
+
+#### Breaking changes
+
+- Split from tantivy: Tantivy is not compatible with this approach.
+- No full entity graph implementation built in
+- No advanced query support
+
+
+### Replication
+
+Each node will have a commit log to track changes through time,
+communicate changes and track replication to read replicas.
+
+- Commit log is append only file which tracks all shard segments
+  and their files. This is then used for replication to read replicas.
+- Writers are responsible for replicating data to readers consistently
+    - 2 phase "commit". First write files everywhere. Second, "turn on" new files.
+    - Having writers responsible for replication allows us to avoid needing
+      to implement something like raft to coordinate read replica state
+- Each node on the system will have a configured set of physical read replicas
+- [OPTIONAL] Readers are responsible for merging segments
+
+
+### Index coordinator
+
+The Index Coordinator is the component responsible for managing the
+IndexNode cluster.
+
+The Index coordinator is the only component that should "know" about
+what shards a Knowledge Box has and how to correctly query or write to them.
+
+Responsibilities:
+- Know writer and reader health
+- Balancing shards across cluster
+    - Place new shards appropriately
+    - Move shards when things are unbalanced
+- Record keeping
+    - Shard stats
+    - Usage stats
+- Index queue consuming, operations
+  - Allow writing to multiple shards on a node at a time(right now, we only write to 1 at a time)
+
+APIs:
+- Create Shard
+- Delete Shard
+- Shard Operation
+    - Search
+    - etc
+
+
+#### Ingest changes
+
+Right now, ingest acts as our "Index Coordinator".
+
+The logic for how to store and query Index Nodes are spread throughout multiple components.
+
+This would be moved to the Index Coordinator.
+
+Changes to ingest/etc:
+- No longer responsible to directly connecting to node writer GPRC service to create new shards
+- No longer responsible for knowing what the best node to create a new shard on
+
+
+### Index Node Writer
+
+- sync changes of commit log to configured read replicas
+- shard create/delete is also part of commit log
+- configured read replicas for Index Node
+
+Write path diagram:
+![Write Path Diagram](./images/005/write-path-ndb-coor.drawio.png)
+
+
+#### Disks
+
+Use best performing persistent disks.
+
+Start with standard SSD persistent disks but move to extreme performance disks
+if performance required.
+
+### Index Node Read replica
+
+Read replicas are "dumb" and are mostly only responsible for serving
+requests from the replicated state.
+
+[OPTIONAL] Readers are responsible for merging segments
+
+Read path diagram:
+![Read Path Diagram](./images/005/read-path-ndb-coor.drawio.png)
```

Comment:
I like the approach search don't have to fetch and merge, but Index Coordinator will have also result merging responsibility. This could become a bottleneck